docs: Generate LogGuide.md contents dynamically

valarnin opened the pull request:
Closes #3380 

CC @quisquous @panicstevenson @MaikoTan

quisquous commented:
What do you think about updating the log guide doc in place, vs having a template copy and a normal copy?

valarnin commented:
I agree that updating in place is better but it doesn't really make sense to do so until everything's moved over. I wanted to get this PR open to get feedback about the code and offer the opportunity for others to help with moving things over, including other languages.

quisquous commented on util/gen_log_guide.ts:
```diff
@@ -0,0 +1,169 @@
+#### Regexes
+
+\`\`\`log
+Network Regex:
```
It's kind of verbose but for consistency, maybe `Network Log Line Regex` and `Network Log Line Examples`?

quisquous commented on util/gen_log_guide.ts:
```diff
@@ -0,0 +1,169 @@
+    },
+    examples: {
+      en: [
+        '00|2021-04-26T14:12:30.0000000-04:00|0839||You change to warrior.|d8c450105ea12854e26eb687579564df',
```
It might be worth having a log line that includes a name too, like Nael rp dialog or something.

valarnin replied:
I don't have any network log versions of Nael quotes handy. I do have this Tsukuyomi quote I have included on the latest commit.

```
00|2021-07-05T18:01:21.0000000-04:00|0044|Tsukuyomi|Oh...it's going to be a long night.|1a81d186fd4d19255f2e01a1694c7607
```

quisquous commented on util/gen_log_guide.ts:
```diff
@@ -0,0 +1,169 @@
+        structureNetworkArray[index] ??= '';
+
+      const structureNetwork = structureNetworkArray.join('|');
+      const structureLogLine = ParseLine.parse(logRepo, structureNetwork);
```
Oh, neat.  I wonder if this will cause problems with some of the `parseFloat`.  I guess most of the positions aren't included in the act log lines, so it will probably work out.

valarnin replied:
Most if not all of the conversion process should be using the original string and not the `parseXxx` value, so that shouldn't be a problem. If there's anywhere that's not using the original string currently, it should be changed to do so.

quisquous commented:
> I agree that updating in place is better but it doesn't really make sense to do so until everything's moved over. I wanted to get this PR open to get feedback about the code and offer the opportunity for others to help with moving things over, including other languages.

I'm not sure I understand what your hesitation is.  It seems to me that you can incrementally replace the blocks of examples and structure with auto-updated ones as they are added.  What are you worries here, or what am I missing in terms of how you are thinking about this?

valarnin commented:
Hmm, that markdownlint error might be a problem if we're going to have links to the structure/regexes/examples entries for each log line type. Should I remove the heading from those?

MaikoTan commented:
Is it make sense to have an enternal JSON file including examples/explanation in various languages, instead of having a extremely long TS file mixed up with code and data.

And by the way, in ja/cn/ko, there are always no any literally "You" in log messages, however, it would be always the name of the current player.
e.g. 
> Taro Satoは「戦士」にチェンジした。
> (Taro Sato changed to Warrior.)(not "You")

I think this should be noted too?

quisquous commented:
> Hmm, that markdownlint error might be a problem if we're going to have links to the structure/regexes/examples entries for each log line type. Should I remove the heading from those?

Oh, it's probably an error because markdown wants to create an anchor for each of those.  Do you think it makes sense to include (redundantly, I know) the log type, e.g. `#### GameLog Structure` or `#### Structure (GameLog)`? Alternatively, feel free to disable that markdown error in the blocks if you think that's gross?

> Is it make sense to have an enternal JSON file including examples/explanation in various languages, instead of having a extremely long TS file mixed up with code and data.

I think for now I could see it as one file or two files as you say, but probably after everything has been added then `lineDocs` will be a very large structure and maybe could live in its own file.

> And by the way, in ja/cn/ko, there are always no any literally "You" in log messages, however, it would be always the name of the current player.

It looks like the examples are by language, so if you want to add examples that are more language-specific, then that should be possible.

valarnin commented:
> Alternatively, feel free to disable that markdown error in the blocks if you think that's gross?

I think disabling the error makes the most sense here. Github's markdown generator already generates the header links as `#structure`, `#structure-1`, `#structure-2`, etc. Since eventually the ToC will be autogenerated (hopefully) I think it's fine to disable.

With regard to splitting the `lineDocs` object to its own file, I agree that it should be moved, but the file will need to be `js`/`ts` or else all the regex strings will have to be hardcoded instead of automatically created based on `NetRegexes`/`Regexes` helpers. I'm not sure where the best place for that file to live would be, or what to call it exactly, so feel free to suggest something.

The final thing I wanted to note was that not every log has a helper function, see in my latest commits:

```ts
  ChangedPlayer: {
    regexes: {
      // @TODO: Add this to NetRegexes/Regexes if we're going to mention them in the log guide?
      network: '',
      logLine: '',
    },
```

Should I add those as part of this PR, or make a new PR to add the missing logs and then rebase once that's merged?

quisquous commented:
> > Alternatively, feel free to disable that markdown error in the blocks if you think that's gross?
> 
> I think disabling the error makes the most sense here. Github's markdown generator already generates the header links as `#structure`, `#structure-1`, `#structure-2`, etc. Since eventually the ToC will be autogenerated (hopefully) I think it's fine to disable.

:+1: with a comment about why you're doing it

> With regard to splitting the `lineDocs` object to its own file, I agree that it should be moved, but the file will need to be `js`/`ts` or else all the regex strings will have to be hardcoded instead of automatically created based on `NetRegexes`/`Regexes` helpers. I'm not sure where the best place for that file to live would be, or what to call it exactly, so feel free to suggest something.

It could just reference the regex helper by name (e.g. `startsUsing` vs `Regexes.startsUsing({ capture: true })`.

> The final thing I wanted to note was that not every log has a helper function, see in my latest commits:
> 
> ```ts
>   ChangedPlayer: {
>     regexes: {
>       // @TODO: Add this to NetRegexes/Regexes if we're going to mention them in the log guide?
>       network: '',
>       logLine: '',
>     },
> ```
> 
> Should I add those as part of this PR, or make a new PR to add the missing logs and then rebase once that's merged?

I think a follow-up PR makes more sense to me.  Maybe better to just have no regexes section at all in those cases?

quisquous commented on docs/LogGuide.md:
```diff
@@ -1,3 +1,6 @@
+<!-- Disable MD004, markdown-magic generates TOC with - instead of * -->
```
MD004's default is "consistent", so you could not disable that by making the other bullet points use `-` instead of `*`.

valarnin replied:
Thanks. I'm not very familiar with markdownlint rules.

valarnin commented:
Just verified in-game that this isn't correct:

```markdown
### 0B: PartyList

Lines are printed, but with blank data.  :sob:
```

0B lines definitely have data in both network and ACT line format.

quisquous commented:
> 0B lines definitely have data in both network and ACT line format.

I wonder if that changed or if I was just always wrong about that.

valarnin commented:
I think this is ready for a full review now.

I may have gone overboard with examples in some cases regarding anything that has a source and a target, I tried to get valid examples that covered all possible combinations.

quisquous commented on docs/LogGuide.md:
```diff
@@ -845,39 +1306,145 @@ Marker Code | Name | Sample Locations | Consistent meaning?
-Unknown?
+This message indicates a floor waymarker was added or deleted.
+
+<!-- AUTO-GENERATED-CONTENT:START (logLines:type=HeadMarker&lang=en) -->
```
I think this is an incorrect type as well.

quisquous commented on docs/LogGuide.md:
```diff
@@ -845,39 +1306,145 @@ Marker Code | Name | Sample Locations | Consistent meaning?
+Network Log Line Examples:
+29|2021-06-10T20:15:15.1000000-04:00|Delete|0|10FF0001|Tini Poutini|4000641D|||50460af5ff3f8ec9ad03e6953d3d1ba9
+29|2021-06-10T20:15:15.1000000-04:00|Add|0|10FF0001|Tini Poutini|400075A9|||0afeb301004fc6d8c7bd5f6cbda9dce7
+29|2021-05-25T22:54:32.5660000-04:00|Add|6|10FF0001|Tini Poutini|10FF0002|Potato Chippy||70a8c8a728d09af83e0a486e8271cc57
```
I think a Delete and Add line are plenty, and we don't need two Adds.

I think also you should probably describe what operation and waymark can be here if you're going to document this log line? (or leave a TODO)?

valarnin replied:
Maybe drop the middle example, so we have an example with ID 0 and an example with ID 6 instead of two 0s?

quisquous commented on docs/LogGuide.md:
```diff
@@ -845,39 +1306,145 @@ Marker Code | Name | Sample Locations | Consistent meaning?
 
-Structure:
-`1E:[ObjectId]:[Target Name] loses the effect of [Status] from [Source Name]`
+<!-- AUTO-GENERATED-CONTENT:START (logLines:type=networkTargetMarker&lang=en) -->
```
wrong type

quisquous commented on docs/LogGuide.md:
```diff
@@ -845,39 +1306,145 @@ Marker Code | Name | Sample Locations | Consistent meaning?
+31|2021-04-26T14:11:35.0000000-04:00|[id]|[data0]|[data1]|[data2]|[data3]
+
+ACT Log Line Structure:
+[14:11:35.0] 1F:[ID]::[data0]:[data1]:[data2]:[data3]
```
Is there an extra colon here, or is this how it looks?

valarnin replied:
Those blank fields should be the player name but that info isn't on the network line. Maybe it makes sense to pre-populate the `LogRepository` instance with the generic combatant info?

quisquous commented on docs/LogGuide.md:
```diff
@@ -925,22 +1492,49 @@ To control aspects of the user interface, the game sends packets called Actor Co
-21:80034E52:8000000D:1601:00:00:00
-21:80037543:80000004:257:00:00:00
+Network Log Line Examples:
+33|2021-04-26T14:17:16.7440000-04:00|80034E5B|8000000C|16|E74|00|00|55e425f79b5cadd724dbc718b86f70b1
```
I think it'd be good to keep the 40000010 wipe line as an example since that's used in a lot of places.

quisquous commented on docs/LogGuide.md:
```diff
@@ -952,69 +1546,120 @@ such as coil turns where there is a zone seal.
-22:105E3321:Tini Poutini:105E3321:Tini Poutini:01
-22:40018065:Twintania:40018065:Twintania:00
+Network Log Line Examples:
+34|2021-04-26T14:19:48.0400000-04:00|4001C51C|Dragon's Head|4001C51C|Dragon's Head|00|a7248aab1da528bf94faf2f4b1728fc3
```
I'd suggest having only two examples here, one 00 and one 01.

quisquous commented on docs/LogGuide.md:
````diff
@@ -952,69 +1546,120 @@ such as coil turns where there is a zone seal.
+
+```log
+Network Log Line Examples:
+35|2021-04-26T17:27:07.0310000-04:00|40003202|Articulated Bit|10FF0001|Tini Poutini|0000|0000|0001|10029769|000F|0000||ad71d456437e6792f68b19dbef9507d5
````
Probably only 2-3 examples would be enough here.

valarnin replied:
One of the examples here was specifically to demonstrate that the names can end up blank. Maybe we should have a note somewhere else in the document stating that the source/target name can be blank on many lines, and they're only guaranteed to be present on, e.g., the 03 line?

quisquous replied:
That might be a good just to note up at the top in general.

quisquous commented on docs/LogGuide.md:
````diff
@@ -1062,40 +1728,94 @@ Examples:
+
+```log
+Network Log Line Examples:
+39|2021-04-26T14:12:29.5060000-04:00|10FF0001|Tini Poutini|121087|121087|10000|10000|0|0|-648.3234|-804.5252|8.570148|1.010669||a0cca3b79b0fd65541e3fa6b11fa386e
````
Probably 2-3 lines here would be ok too.

valarnin replied:
I included the first four lines here to demonstrate that in some cases these come in at 3 second intervals. Maybe it's better explained as plain text?

quisquous replied:
Oh, I definitely missed what that was trying to demonstrate.  Plain text with fewer examples sounds better to me.

quisquous commented on util/gen_log_guide.ts:
```diff
@@ -0,0 +1,503 @@
+      ],
+    },
+  },
+  limitBreak: {
```
Why is `limitBreak` lowercase vs other types?

valarnin replied:
The keys on this object line up with the keys on `logDefinitions`. Only the keys on `logDefinitions` that have a corresponding `NetRegexes` function are ProperCase.

quisquous replied:
Hmm, ok, then `logDefinitions` is inconsistent.  That doesn't need to be fixed up in this PR then, but I think consistent capitalization would be better both here and there.

valarnin replied:
Yeah, that was mentioned in #3377, they were previously all camelCase but I adjusted some of them there to match the expected format downstream in triggers to avoid having to update 3000-something triggers.

quisquous replied:
Fixed in #3386.

valarnin replied:
Okay. Do you want me to rebase after that's merged, or are you going to resolve conflicts as part of merging this one?

quisquous replied:
I can rebase my patch if that's easier for you.  I wasn't trying to create extra work for you.  :)

valarnin replied:
I'll go ahead and rebase this one after yours is merged, since yours is the simpler change.

quisquous commented on util/gen_log_guide.ts:
```diff
@@ -0,0 +1,503 @@
+const lineDocs: LineDocs = {
+  GameLog: {
+    regexes: {
+      network: NetRegexes.gameLog({ capture: true }).source,
```
What do you think about just having `regex: 'gameLog'` here, and not repeating `{ capture: true }).source` everywhere?

valarnin replied:
I thought about that originally but decided against it so that regexes could be manually specified if needed. There's also the case of `nameToggle` where the key exists on one but not the other.

quisquous commented on util/gen_log_guide.ts:
```diff
@@ -0,0 +1,503 @@
+      const examplesNetwork = lineDoc.examples[language]?.join('\n') ?? '';
+      const examplesLogLine = lineDoc.examples[language]?.map((e) => {
+        const line = ParseLine.parse(logRepo, e);
+        return line?.properCaseConvertedLine ?? line?.convertedLine;
```
What do you think about line being undefined printing an error an exiting here.  I think that would be unexpected, right?

valarnin replied:
Since we're controlling the input examples, this would be more of an UnreachableCode situation?

quisquous commented on util/gen_log_guide.ts:
```diff
@@ -0,0 +1,503 @@
+import LogRepository from '../ui/raidboss/emulator/data/network_log_converter/LogRepository';
+import ParseLine from '../ui/raidboss/emulator/data/network_log_converter/ParseLine';
+
+type LangObject<T> = { en: T } & { [lang in NonEnLang]?: T };
```
I think this is a `LocaleObject<T>` from `trigger.d.ts`?

valarnin replied:
Thank you. I knew we had a type for this already but I kept searching for Lang, instead of Locale.

quisquous commented on util/gen_log_guide.ts:
```diff
@@ -0,0 +1,503 @@
+    let exitCode = 0;
+    for (const lang of languages) {
+      // Only check this lang if it was converted/base file existed
+      if (
+        !(
+          output.some((o) =>
+            RegExp(('[^\\w]' + lang + '[^\\w]')).exec(o.originalPath.toLowerCase())
+          ) ||
+          (lang === 'en' && !output.some((o) => o.originalPath.includes(enLogGuidePath)))
+        )
+      )
+        continue;
```
As a simplification suggestion, what about ignoring any language that has no mappedLogLines instead?

valarnin replied:
I think maybe it makes sense to define the potential error states and what action to take (warning or error) for each state, then adjust the logic to match?

Here are the potential error states as I see them:

1. No language-specific markdown file found (e.g. no `docs/de/LogGuide.md` file)
  i. Currently ignored
2. Markdown file doesn't have every defined type in `lineDocs`
  i. Currently errors
3. Markdown file attempts to reference a type not defined in `lineDocs`
  i. Currently errors
4. Markdown file attempts to reference an invalid language
  i. Currently errors

If this is the behavior we want for these error states, then adjusting the logic as suggested breaks `2` since a `docs/de/LogGuide.md` file with no mapped lines referenced would not error.

quisquous replied:
I guess I'm saying that it is unlikely that a markdown file would exist and have no translations.  Even though ignoring no mapped lines breaks a strict interpretation of (2), in practice I think the only case that we'll ever see is when translations exist and one is forgotten.  It's a much simpler condition to check correctly, which is why I suggest it, rather than future me trying to re-puzzle over this conditional.

valarnin replied:
I have updated the logic used for checking for errors. The output looks like this if a blank LogGuide.md file is created in the zh-CN folder:

```
✔ docs\LogGuide.md Updated
 Transforms run
  ⁕ logLines:type=GameLog&lang=en
  ⁕ logLines:type=ChangeZone&lang=en
  ⁕ logLines:type=ChangedPlayer&lang=en
  ⁕ logLines:type=AddedCombatant&lang=en
  ⁕ logLines:type=RemovedCombatant&lang=en
  ⁕ logLines:type=partyList&lang=en
  ⁕ logLines:type=PlayerStats&lang=en
  ⁕ logLines:type=StartsUsing&lang=en
  ⁕ logLines:type=Ability&lang=en
  ⁕ logLines:type=networkCancelAbility&lang=en
  ⁕ logLines:type=networkDoT&lang=en
  ⁕ logLines:type=WasDefeated&lang=en
  ⁕ logLines:type=GainsEffect&lang=en
  ⁕ logLines:type=HeadMarker&lang=en
  ⁕ logLines:type=networkRaidMarker&lang=en
  ⁕ logLines:type=networkTargetMarker&lang=en
  ⁕ logLines:type=LosesEffect&lang=en
  ⁕ logLines:type=networkGauge&lang=en
  ⁕ logLines:type=ActorControl&lang=en
  ⁕ logLines:type=NameToggle&lang=en
  ⁕ logLines:type=Tether&lang=en
  ⁕ logLines:type=limitBreak&lang=en
  ⁕ logLines:type=StatusEffect&lang=en
  ⁕ logLines:type=networkUpdateHP&lang=en
  ⁕ TOC

Language cn is missing LogGuide doc entry for type GameLog
Language cn is missing LogGuide doc entry for type ChangeZone
Language cn is missing LogGuide doc entry for type ChangedPlayer
Language cn is missing LogGuide doc entry for type AddedCombatant
Language cn is missing LogGuide doc entry for type RemovedCombatant
Language cn is missing LogGuide doc entry for type partyList
Language cn is missing LogGuide doc entry for type PlayerStats
Language cn is missing LogGuide doc entry for type StartsUsing
Language cn is missing LogGuide doc entry for type Ability
Language cn is missing LogGuide doc entry for type networkCancelAbility
Language cn is missing LogGuide doc entry for type networkDoT
Language cn is missing LogGuide doc entry for type WasDefeated
Language cn is missing LogGuide doc entry for type GainsEffect
Language cn is missing LogGuide doc entry for type HeadMarker
Language cn is missing LogGuide doc entry for type networkRaidMarker
Language cn is missing LogGuide doc entry for type networkTargetMarker
Language cn is missing LogGuide doc entry for type LosesEffect
Language cn is missing LogGuide doc entry for type networkGauge
Language cn is missing LogGuide doc entry for type ActorControl
Language cn is missing LogGuide doc entry for type NameToggle
Language cn is missing LogGuide doc entry for type Tether
Language cn is missing LogGuide doc entry for type limitBreak
Language cn is missing LogGuide doc entry for type StatusEffect
Language cn is missing LogGuide doc entry for type networkUpdateHP
npm ERR! code ELIFECYCLE
npm ERR! errno 1
```

Hopefully this logic makes more sense?

quisquous replied:
Sounds good, thanks!

quisquous commented on util/gen_log_guide.ts:
```diff
@@ -0,0 +1,503 @@
+      const structureLog = structureLogLine?.properCaseConvertedLine ??
+        structureLogLine?.convertedLine ?? '';
+
+      const examplesNetwork = lineDoc.examples[language]?.join('\n') ?? '';
+      const examplesLogLine = lineDoc.examples[language]?.map((e) => {
```
I think this `[language]` lookup here should behave like cactbot translation functions, where it defaults to English if the language itself doesn't provide different examples.  I suspect in the vast majority of cases there will not need to be translated log lines.

quisquous reviewed (commented):
Thanks for all your work here.  I think this is all really good stuff, will make it really easy to add in new lines, and adds so much extra valuable reference material for others to look at.

> I may have gone overboard with examples in some cases regarding anything that has a source and a target, I tried to get valid examples that covered all possible combinations.

Yeah, I think that's my biggest feedback here.  I think the examples are there to just help elucidate the structure.  I think in the vast majority of cases there should just be 2 (or maybe 3) examples at the most.  If there were more before, it's because there were fewer sections.  I think the structure / regex / examples gets pretty long, so limiting examples seems good.

valarnin commented on docs/LogGuide.md:
```diff
@@ -328,65 +426,135 @@ and does not consistently show up first.
-03:40123456:Added new combatant Pagos Deepeye.  Job: N/A Level: 70 Max HP: 348652 Max MP: 12000 Pos: (-720.9337,90.80706,-679.6056).
-03:10987654:Added new combatant Tater Tot (Jenova).  Job: 28 Level: 70 Max HP: 39835 Max MP: 16461 Pos: (-143.9604,168.5795,-4.999999).
+Network Log Line Structure:
+03|2021-04-26T14:11:35.0000000-04:00|[id]|[name]|[job]|[level]|[ownerId]|[worldId]|[world]|[npcNameId]|[npcBaseId]|[currentHp]|[hp]|[currentMp]|[mp]|||[x]|[y]|[z]|[heading]
```
`[?]` on unhandled fields works.

`[timestamp]` would require doing a string replace on the ACT log line version after the conversion, which is a bit ugly.

valarnin replied:
Alright, this is done as of the latest pushed commit. As stated, `[timestamp]` is a bit ugly, so I'm still not sure if there's a better way to handle it.

quisquous replied:
Looks fine to me, thanks!

valarnin commented on docs/LogGuide.md:
```diff
@@ -548,8 +834,47 @@ If one person runs the fireball out and it only hits them, then it is type `15`
+^(?<type>(?:2[12]))\|(?<timestamp>(?:[^|]*))\|(?<sourceId>(?:[^|]*))\|(?<source>(?:[^|]*))\|(?<id>(?:[^|]*))\|(?<ability>(?:[^|]*))\|(?<targetId>(?:[^|]*))\|(?<target>(?:[^|]*))\|(?<flags>(?:[^|]*))\|(?<damage>(?:[^|]*))\|(?:[^|]*\|){14}(?<targetCurrentHp>(?:[^|]*))\|(?<targetMaxHp>(?:[^|]*))\|(?<targetCurrentMp>(?:[^|]*))\|(?<targetMaxMp>(?:[^|]*))\|(?:[^|]*\|){2}(?<targetX>(?:[^|]*))\|(?<target ...
+
+ACT Log Line Regex:
+(?<timestamp>(?:^.{14})) 1[56]:(?<sourceId>(?:[0-9A-F]{8})):(?<source>(?:[^:]*?)):(?<id>(?:[0-9A-Fa-f]{1,8})):(?<ability>(?:[^:]*?)):(?<targetId>(?:[0-9A-F]{8})):(?<target>(?:[^:]*?)):(?<flags>(?:[^:]*?)):(?<flag0>(?:[^:]*?)):(?<flag1>(?:[^:]*?)):(?<flag2>(?:[^:]*?)):(?<flag3>(?:[^:]*?)):(?<flag4>(?:[^:]*?)):(?<flag5>(?:[^:]*?)):(?<flag6>(?:[^:]*?)):(?<flag7>(?:[^:]*?)):(?<flag8>(?:[^:]*?)):(?<fl ...
```
Yeah, they're all `Regexes.maybeCapture`'d in Regexes

valarnin commented on docs/LogGuide.md:
```diff
@@ -712,48 +1055,134 @@ Instead, if a boss has 20 dots applied to it,
+[14:17:42.890] 1A:4000016E:4001c490 gains the effect of Combust II from Potato Chippy for 30.00 Seconds.
+[14:14:09.165] 1A:4000016E:10ff0002 gains the effect of Wildfire from Potato Chippy for 10.00 Seconds.
+[14:36:17.251] 1A:4002E66C:Speckled Spider gains the effect of Venomous Bite from  for 30.00 Seconds.
+[21:57:07.911] 1A:40003D9D: gains the effect of Damage Down from Eden's Promise for 30.00 Seconds.
```
Line 1180 has source name missing, line 1181 has target name missing. I think this is correct?

```log
Network line:
26|2021-07-02T21:57:07.9110000-04:00|82c|Damage Down|30.00|40003D9F|Eden's Promise|40003D9D||00|148000|63981880||86ff6bf4cfdd68491274fce1db5677e8

ACT line:
[21:57:07.911] 1A:40003D9D: gains the effect of Damage Down from Eden's Promise for 30.00 Seconds.
```

quisquous replied:
Oh, sorry, I misread that, then.

MaikoTan commented:
Maybe to add a little "return to top" link in every section bottom make sense? I think if readers are able to jump between sections easily, it shouldn't be very painful to have a bit longer examples or something.

valarnin commented:
> Maybe to add a little "return to top" link in every section bottom make sense? I think if readers are able to jump between sections easily, it shouldn't be very painful to have a bit longer examples or something.

Easy enough for me to add, if quisquous agrees.

quisquous commented:
> > Maybe to add a little "return to top" link in every section bottom make sense? I think if readers are able to jump between sections easily, it shouldn't be very painful to have a bit longer examples or something.
> 
> Easy enough for me to add, if quisquous agrees.

I can go either way.  I lean no, both because I don't think we need longer examples, but also because the home key / scrollbar / back button if somebody clicked on a heading bring people back to the top.  However, that's just a weak no, and if you both think it should be added, then go for it.  Up to you.

valarnin commented on docs/LogGuide.md:
```diff
@@ -712,48 +1055,134 @@ Instead, if a boss has 20 dots applied to it,
+24|2021-04-26T14:11:35.0000000-04:00|[id]|[name]|[which]|[effectId]|[damage]|[currentHp]|[maxHp]|[currentMp]|[maxMp]|||[x]|[y]|[z]|[heading]
+
+ACT Log Line Structure:
+[14:11:35.0] 18:24 Tick on [Name] for NaN damage.
```
This is now fixed. I'm not really too happy with the fix but I couldn't see any other way to handle this case.

valarnin commented:
Alright, I think I've addressed everything from the big review left by quisquous.

Rebased to main to pick up the changes to case from the other PR.

Are there still enough examples that you think there needs to be a link to top now, @MaikoTan?

valarnin commented on docs/LogGuide.md:
```diff
@@ -845,39 +1280,165 @@ Marker Code | Name | Sample Locations | Consistent meaning?
 ### 1C: NetworkRaidMarker
 
-Unknown?
+This message indicates a floor waymarker was added or deleted.
```
I put these in backwards due to autopilot during a work meeting 🙃 

28 = floor marker
29 = combatant marker

valarnin commented on docs/LogGuide.md:
```diff
@@ -400,16 +560,42 @@ This message is sent when an object is removed from the scene, either because
+
+ACT Log Line Examples:
+[02:01:27.548] 04:10FF0001:Removing combatant Tini Poutini. Max MP: 10000. Pos: (-66.24337,-292.0904,20.06466)
+[00:37:36.74] 04:4000B39C:Removing combatant Petrosphere. Max MP: 10000. Pos: (-16.00671,-0.01531982,0)
```
I've added padding out to 3 decimal places, but I'm not sure if that's accurate for a `.0` case. I'd have to look at the FFXIV parser plugin's decomp or try to track down a line and manage to feed it through again.

I'm not sure if it's worth the effort?

quisquous replied:
I am 90% sure that it's `.000` as you have it written having looked a lot of log lines over time.  This looks good to me.

valarnin commented on docs/LogGuide.md:
```diff
@@ -328,65 +428,125 @@ and does not consistently show up first.
+03|[timestamp]|[id]|[name]|[job]|[level]|[ownerId]|[worldId]|[world]|[npcNameId]|[npcBaseId]|[currentHp]|[hp]|[currentMp]|[mp]|[?]|[?]|[x]|[y]|[z]|[heading]
+
+ACT Log Line Structure:
+[timestamp] 03:[ID]:Added new combatant [name]([world]).  Job: NONE Level: [level] Max HP: [hp] Max MP: [mp] Pos: ([x],[y],[z]) ([npcNameId][npcBaseId]).
```
This can't be fixed in the same way as the damage issue unfortunately, since having a blank job is a valid case that defaults to NONE for e.g. enemy combatants. I think this one will have to be a special case in the generator instead?

quisquous replied:
Did you want to solve this before I merge this?

valarnin replied:
Solving this would basically be something like this in the transform function:

```ts
if (type === 'AddedCombatant')
  structureLog = structureLog.replace(/Job: NONE/, 'Job: [job]')
```

Which is pretty ugly. If you think that's an okay fix then sure, I can add it.

quisquous replied:
I don't know.  I think having the log structure be clearer is better than having a clean script, so go for it?  Agreed that it's pretty ugly but I can't see another way to do it.

valarnin commented:
This should be good to merge now.
